### ISSUE TITLE
Error running it after server.close()

### ISSUE THREAD

**Issue by mercmobily** [myndzi/wtfnode#14]
If you run `wtf.dump()` after closing the server, you *might* get (it depends on timing):

    - Servers:
    /var/www/node-apps/gigsnet/node_modules/wtfnode/index.js:430
                console.log('  - %s:%s (%s)', a.address, a.port, type);
    TypeError: Cannot read property 'address' of null

May I suggest changing:

            console.log('  - %s:%s (%s)', a.address, a.port, type);
            
With:

            if (a) console.log('  - %s:%s (%s)', a.address, a.port, type);
            else console.log("No address available")

...?

**Comment by myndzi** [comment]
Seems reasonable! I emphasize that this entire module is a huge hack, but I'm glad people get use out of it, so I'm happy to fix up problems as they arise. I fear I didn't write it to be very "release-worthy" at the outset,  which is why there's stuff like this laying around...

**Comment by mercmobily** [comment]
Honest hat on: your module saved my life! I was going insane trying to figure out where on earth the callback was -- or even any trace info -- for timeouts, as I had one lying around.
I am trying to decide whether I want to leave this running as production 0 -- which I probably will.
I am a "graceful stop" nazi, and am using this to check what's left in the queue after SIGTERM. Since at that point the process doesn't even listen to the http port, it doesn't even matter if it just lies around.

Anyhow... do you want me to do a PR? It's probably much easier for you to just change it...

THANK YOU!

**Comment by myndzi** [comment]
I don't mind changing it, but I probably won't get to it until tomorrow. I'm glad the module helped you! That sort of frustration is exactly why I wrote it :) I don't recommend running it in production, it does Bad Things with unpredictable effects, not the least of which is wrapping/getting involved in timers and event emitters, the latter of which are used extensively in Node's guts.

**Comment by myndzi** [comment]
Just kidding, I did it anyway ;)

**Comment by mercmobily** [comment]
Thanks!
Hummm not having it in production means that I won't have valuable information when I need it the most... what a pain! I guess for production I will need to just deal with the list of entries...

I don't suppose we could lobby the node guys to add info to the event loop so that NO hacks in this module are necessary?

**Comment by myndzi** [comment]
They're definitely making it harder to do the kind of hooking I'm doing, as things move to `internal/`, but while I think some sort of public API to get the open handles information more easily is reasonable, I'm not sure that asking them to also instrument everything to provide callsite information and such (in other words, useful information to tell you what part of YOUR code is related to the problem) is useful or likely to be accepted.

The best use here is with an application for which the code paths that may cause dangling handles are tested, and to use `wtfnode` on the fly if you can't figure out why things are hanging. If you have something related to the operation of your application that causes dangling handles, I'd say that's within your own power to keep tabs on without resorting to this sort of thing.

In other words, `wtfnode` should help you find _flaws in application logic_ that don't clean up correctly, not _runtime state_. If you tell me a bit more about what's going on, I may be able to help you come up with alternate ideas. That said, I'm not saying you _can't_ use it in production, I'm just saying that if you do, you may be subject to weird, unexpected problems that arise as node's own code changes in ways that may not play well with what `wtfnode` does. _I_ don't think that's a worthwhile risk, but for your use case, it might be.

**Comment by myndzi** [comment]
To elaborate a little more, you stated that it "doesn't matter" if it (wtfnode?) lies around. I think this betrays a misunderstanding of what the module is actually doing.

When it's loaded, `wtfnode` fundamentally interferes with core components of Node's operations in order to get the callsite information. It does this by monkey patching private methods on core Node APIs. Whenever you set a timer, or whenever anything creates an event handler (this includes things like `server.listen(callback)`), there's an extra layer added between the call site and the code that deals with those things.

That extra layer is not, in my opinion as its author, suitably guaranteed to be problem-free. It doesn't need to be, for the troubleshooting use-case -- the worst that can happen there is it breaks or doesn't answer your question -- but for running production code, I'd *want* it to be as battle-tested as Node itself is. I just don't have the resources or thorough knowledge of everything I'd need to know about to make that the case.

**Comment by mercmobily** [comment]
Oh no I get you. By "it" I meant "The app process that was SIGTERM'ed". Since the process is no longer listening to a socket, I can run the NEW server while that one is still going, if need be, finishing off whatever it was doing.
Also... I know what you mean. WTF makes a lot of sense in debugging mode. What bothers me is that there are just so many modules that might set timers, intervals, etc., and you might not be fully aware of what's going on.
My battle was in the mongo driver's field: the monitorProcess() function in `lib/topologies/server.js` will set a self-renewing timeout every 5 seconds. THIS is the one thing that, without wtfnode, I would have never found.

**Comment by myndzi** [comment]
Ahh, I was trying to relate your comment to why you'd run `wtfnode` in prod. Problems with a dependency are the worst. If it was possible, I'd provide a way to hook things on demand at runtime, but it really isn't. :\

**Comment by mercmobily** [comment]
I know... hence my idea of lobbying the node guys till they realise that this module should be able to do what it does WITHOUT any hacks. It's a great use-case, and it makes it easy to see when it's "fixed"... if you know what I mean.

Would you feel like writing a detailed list of:

* The info you need
* The hack you have to execute to get it
* The way it "should" be available in node

...?
That would be pretty much "it" -- I'd be happy to create an issue in node, referencing this issue)

**Comment by myndzi** [comment]
The gist is simple. I need a public API to get the open handles, and for everything that creates a handle that can keep node open, I need:
- A way to gather what part of the application code created it and
- A way to link the handle back to that information

For the first item, a full stack trace is acceptable and more generic.

The problems I have to solve with `wtfnode` are twofold:
1. Hook invocations so I can capture this data somewhere and
2. Create a way to work my way back to obtaining this data given a handle reference

The linkage from handle -> data is not always easy, and more things are being hidden away to places where I can't monkey patch them under `internal/`, or wrapped in scopes, in such a way that the only solution is to apply a broader or less targeted monkey patch at a higher level.

I'm not really sure what the best way to solve these things in Node core itself would be, or how it would affect performance and memory. There is a semantic difference between doing something like "bind an event listener", which is where and when I need #1, and "create a handle that may keep Node running", which is where and when I need #2. The way to work from #2 to #1 depends on the particulars of the handle in question.

**Comment by mercmobily** [comment]
But... shouldn't it be a matter of asking them to make sure that:

     process._getActiveHandles()

Returns everything we need?
That would be:

* For Timers: the callback and the stack trace of when it was set
* For Sockets: when it was set
...more?

I don't think it makes sense to have this info in the event loop itself --
so, _getActiveHandles should return the event loop *enriched* with extra
debug info...

Merc.

**Comment by myndzi** [comment]
You'll note that `_getActiveHandles` is prefixed with a `_`, which hints that it's not a public API. In itself, it's not in a position to sustain requests for changes intended for external consumers, which means that this request would inherently be a request for a public API. Separately from that, *instrumenting* everything to provide the information required doesn't seem to be a currently existing pattern, and might well bring with it various complications affecting other areas of the code. I don't think it's necessarily a bad idea, but it's definitely not a small, well-contained change.

**Comment by mercmobily** [comment]
Good point... OK let's forget about it for now, sorry about the noise :D

**Comment by myndzi** [comment]
No worries. I think it could be good to provoke a discussion about serving this need better, I just don't expect that it's straightforward or necessarily likely to happen. I encourage you to raise it, however!
